File: ruby/lib/task_list.rb
require_relative 'task'

class TaskList
  QUIT = 'quit'

  def initialize(input, output)
    @input = input
    @output = output

    @tasks = {}
  end

  def run
    while true
      @output.print('> ')
      @output.flush

      command = @input.readline.strip
      break if command == QUIT

      execute(command)
    end
  end

private

  def execute(command_line)
    command, rest = command_line.split(/ /, 2)
    case command
    when 'show'
      show
    when 'add'
      add rest
    when 'check'
      check rest
    when 'uncheck'
      uncheck rest
    when 'help'
      help
    else
      error command
    end
  end

  def show
    @tasks.each do |project_name, project_tasks|
      @output.puts project_name
      project_tasks.each do |task|
        @output.printf("  [%c] %d: %s\n", (task.done? ? 'x' : ' '), task.id, task.description)
      end
      @output.puts
    end
  end

  def add(command_line)
    subcommand, rest = command_line.split(/ /, 2)
    if subcommand == 'project'
      add_project rest
    elsif subcommand == 'task'
      project, description = rest.split(/ /, 2)
      add_task project, description
    end
  end

  def add_project(name)
    @tasks[name] = []
  end

  def add_task(project, description)
    project_tasks = @tasks[project]
    if project_tasks.nil?
      @output.printf("Could not find a project with the name \"%s\".\n", project)
      return
    end
    project_tasks << Task.new(next_id, description, false)
  end

  def check(id_string)
    set_done(id_string, true)
  end

  def uncheck(id_string)
    set_done(id_string, false)
  end

  def set_done(id_string, done)
    id = id_string.to_i
    task = @tasks.collect { |project_name, project_tasks|
      project_tasks.find { |t| t.id == id }
    }.reject(&:nil?).first

    if task.nil?
      @output.printf("Could not find a task with an ID of %d.\n", id)
      return
    end

    task.done = done
  end

  def help
    @output.puts('Commands:')
    @output.puts('  show')
    @output.puts('  add project <project name>')
    @output.puts('  add task <project name> <task description>')
    @output.puts('  check <task ID>')
    @output.puts('  uncheck <task ID>')
    @output.puts()
  end

  def error(command)
    @output.printf("I don't know what the command \"%s\" is.\n", command)
  end

  def next_id
    @last_id ||= 0
    @last_id += 1
    @last_id
  end
end

if __FILE__ == $0
  TaskList.new($stdin, $stdout).run
end

File: typescript/src/task_list.ts
import * as readline from 'readline';
import * as util from 'util';
import { Task } from './task';

enum Commands {
    SHOW = 'show',
    ADD_PROJECT = 'add project',
    ADD_TASK = 'add task',
    CHECK = 'check',
    UNCHECK = 'uncheck',
    HELP = 'help',
    QUIT = "QUIT"
}

export class TaskList {
    private readonly readlineInterface;
    private readonly tasks: Record<string, Task[]> = {};
    private lastId = 0;

    constructor(private readonly reader: NodeJS.ReadableStream, private readonly writer: NodeJS.WritableStream) {
        this.readlineInterface = readline.createInterface({
            terminal: false,
            input: reader,
            output: writer,
        });

        this.setupReadline();
    }

    private setupReadline() {
        this.readlineInterface.setPrompt('> ');
        this.readlineInterface.on('line', this.handleCommand.bind(this));
        this.readlineInterface.on('close', () => this.writer.end());
    }

    private handleCommand(commandLine: string) {
        const [command, rest] = this.splitFirstSpace(commandLine);

        switch (command) {
            case Commands.QUIT:
                this.readlineInterface.close();
                break;
            case Commands.SHOW:
                this.show();
                break;
            case Commands.ADD_PROJECT:
                this.addProject(rest);
                break;
            case Commands.ADD_TASK:
                this.addTask(rest);
                break;
            case Commands.CHECK:
                this.setDone(rest, true);
                break;
            case Commands.UNCHECK:
                this.setDone(rest, false);
                break;
            case Commands.HELP:
                this.help();
                break;
            default:
                this.error(command);
                break;
        }

        this.readlineInterface.prompt();
    }

    private splitFirstSpace(s: string): [string, string] {
        const pos = s.indexOf(' ');
        if (pos === -1) {
            return [s, ''];
        }
        return [s.substr(0, pos), s.substr(pos + 1)];
    }

    private show() {
        for (const [project, taskList] of Object.entries(this.tasks)) {
            this.println(project);
            taskList.forEach((task) => {
                this.println(util.format('    [%s] %d: %s', task.done ? 'x' : ' ', task.id, task.description));
            });
            this.println('');
        }
    }

    private addProject(name: string) {
        this.tasks[name] = [];
    }

    private addTask(commandLine: string) {
        const [project, description] = this.splitFirstSpace(commandLine);
        const projectTasks = this.tasks[project];

        if (!projectTasks) {
            this.println(util.format('Could not find a project with the name "%s".', project));
            return;
        }

        projectTasks.push(new Task(this.nextId(), description, false));
    }

    private setDone(idString: string, done: boolean) {
        const id = parseInt(idString, 10);

        for (const taskList of Object.values(this.tasks)) {
            const task = taskList.find((t) => t.id === id);
            if (task) {
                task.done = done;
                return;
            }
        }

        this.println(util.format('Could not find a task with an ID of %d.', id));
    }

    private help() {
        this.println('Commands:');
        this.println('  show');
        this.println('  add project <project name>');
        this.println('  add task <project name> <task description>');
        this.println('  check <task ID>');
        this.println('  uncheck <task ID>');
        this.println('');
    }

    private error(command: string) {
        this.println(`I don't know what the command "${command}" is.`);
    }

    private println(ln: string) {
        this.readlineInterface.output.write(ln);
        this.readlineInterface.output.write('\n');
    }

    private nextId(): number {
        return ++this.lastId;
    }

    run() {
        this.readlineInterface.prompt();
    }
}

if (require.main === module) {
    new TaskList(process.stdin, process.stdout).run();
}
